src/levanter/data/loader.py

```diff
@@ -31,23 +33,60 @@
 logger = logging.getLogger(__name__)
 
 
+def local_indices_for_bs_and_sharding(sharding: jax.sharding.Sharding, batch_size: int) -> dict[jax.Device, range]:
+    """Return the local indices for each device for a given batch size and sharding."""
+    local_indices_map = sharding.addressable_devices_indices_map((batch_size,))
+    # we just want all the indices
+    local_device_indices: dict[jax.Device, range] = {
+        device: range(*idx[0].indices(batch_size)) for device, idx in local_indices_map.items() if idx is not None
+    }
+    return local_device_indices
+
+
+# NOTE: In general there are a lot of different indices flying around. Here's a quick guide:
+# - `step` or `batch_number` or `bn` is the training step or batch number
+# - `global` indices refer to the index into the datastore
+# - `local` indices refer to the index into the batch. These start at 0 for every batch and go up to the batch size.
+
+# That is, for each batch, we have:
+# global_offset = global_offset_at_step(bn)
+# batch_size = batch_size_at_step(bn)
+# local_data_to_global_map = {i: global_offset + i for i in range(batch_size)}
+
+# But note that we only load the data we need for our local devices, so we only load a subset of the data for each batch.
+
+
+@dataclass
+class _Batch(Generic[Ex]):
+    index: int
+    global_data_offset: int
+    global_size: int
+    data_by_local_index: dict[int, Ex]
+
+
 class DataLoader(Iterable[Ex]):
     def __init__(
         self,
-        Batch: hax.Axis,
         data: AsyncDataset[Ex],
-        max_buffered_batches: Optional[int],
+        batch_axis_name: str | hax.Axis,
+        batch_size: int | IntSchedule | None = None,
+        *,
+        max_buffered_batches: Optional[int] = 64,
         mesh: Mesh,
         axis_resources: Optional[ResourceMapping],
-        # this is set heuristically for the typical tokenseqdataset we use. Should probably tune
         prefetch_size: int = 32,
     ):
         """
+        Batch- and NamedArray-aware data loader. This class works with an [AsyncDataset][], a Mesh,
+        and a resource mapping to load data in a way that is aware of the batch axis and the sharding
+        of the data. In general, each machine only loads the data that it needs to process.
 
-        TODO: document this
+        (At the moment, it's not smart enough to load parts of examples if examples are sharded across machines,
+        but whole examples are handled correctly.)
 
         Args:
-            Batch (hax.Axis): The batch axis
+            batch_axis_name (str | Axis): The name of the batch axis
+            batch_size (int | IntSchedule | None): The size of the batch or a schedule for the size of the batch
```

Comment:
Hmmm, I think this risks being confusing since batch_axis_name sometimes sneakily contains the batch_size too.

Is it to un-haliax-y to only support `batch_axis_name (str)` and `batch_size (int | IntSchedule)`?

Reply:
eh, i thought about this. it was kind of an attempt at backwards compat but I guess there's no reason

Reply:
i'll just commit

Reply:
err, int | IntSchedule | Axis